### DISCUSSION TITLE
Performance issue on outgoing requests

### DISCUSSION THREAD

**Discussion by rehmatworks** [fastapi/fastapi#6664] [Questions] [answered]
We have an API developed using DRF but recently we switched to FastAPI. We are loving it a lot. This is one of the greatest software that we have ever come across.

Now getting back to the point, our upcoming version of API is built using FastAPI. When we deploy it to a Docker Swarm cluster, regular endpoints (that just return JSON responses without making any computation) perform really great. In our small cluster, we achieved 40K+ requests per second in `wrk` benchmarks.

But when we call the endpoints where we are making an outgoing request, the performance dramatically decreases. We barely achieve 100 requests per second in our benchmarking tests.

The endpoints where the performance issue is present make outgoing requests and we return data in real-time. We are using a proxy service too so latency issues are there. But the performance difference between the normal endpoints and the ones that get external data and return to the user is really really huge.

We have tried using Gunicorn + Uvicorn worker and Uvicorn alone too. We also tried applying an API gateway like Kong but the performance issue remains there. On the nodes where the API is running CPU usage rises while the API gateway doesn't go under any load.

Is this a CPU-intensive case or threads are locked when external requests response is awaited? We are using httpx and each API call results in a single external request. Any help will be appreciated greatly.

**Comment by ArcLightSlavik** [comment]
Could you provide examples/more information?

a. Are you **sure** the service ur calling isn't the bottleneck, how many services are you calling in each request to FastAPI?
b. How many gunicorn/uvicorn workers?
c. Have you tried a different outgoing service / only calling once?
d. Could you give more info / data about the CPU usage you're getting?
e. Did you run other benchmarks with other endpoints that require slightly less computation?
f. Have you tried other benchmarking tools?

With the information I have right now I can't really see how the benchmark would be that wrong. Your comparing the computation endpoint to a simple json response which will always be insanely fast, you need to test other endpoints that do different actions.

**Comment by rehmatworks** [comment]
Hi @ArcLightSlavik, thank you for replying.

a. The services that I am calling take up to 5 seconds to respond
b. (2 * 2) + 1 per core (And I tested with 6 nodes each with 4 cores, so 120 Gunicorn workers)
c. Yep, same poor performance with all. And the service is called once per request
d. CPU usage across the cluster reaches ~100% before connections start timing out
e. I have not tested other endpoints yet. What I have tested do no intense computation and they just return a predefined JSON response
f. Not yet, I'll try and I'll post the output here

**Update**: I tried Apache Bench and the results are the same.

**Comment by STeveShary** [comment]
@rehmatworks Some thoughts with your performance:
- The "(2 * 2) + 1 per core" is a general thought.  It really might matter based on if you are using a cloud server that is sharing with others and the CPU/core layout.  It might depend if you have alot of JSON transformations, business logic....
- Try to adjust the different parameters and perform a load testing with something like: https://locust.io/. I use this personally and it really helps tune things correctly.
- Are you hitting an HTTPS endpoint with your external call?   You may find that you client is spending all of your CPU establishing TLS connections.  Look at a pooling option so you don't pay for that setup with every connection.
- What HTTP client are you using?  I use AIOHttp and have to tune our clients to get optimal performance: https://docs.aiohttp.org/en/stable/client_advanced.html#connectors
- Lastly, try to add some telemetry with an APM tool: New Relic, AppDynamics, Dynatrace, or add some low level logging so you can understand which functions and layers are your bottlenecks.

It sounds like you are getting good performance with a simple use case.  Calling external servicing making it harder and you have to spend the time to get "optimal" performance.

**Comment by rehmatworks** [comment, marked as answer]
@STeveShary Thank you so much for replying. Looks like you have pointed me in the right direction. I am modifying my code and I'll run some benchmarking tests again.

- I'll give locust.io a try
- Yes, the endpoints are HTTPS. I am now using HTTP endpoints as well as I am trying pooling
- I am using [HTTPX](https://python-httpx.org)
- I am using Sentry but haven't looked in the logs for this particular issue yet

Thank you for all your help. I'll be running benchmarking tests again and I'll share the results.

Update:
Connection pooling does work and I achieved 8K+ requests per second and the results were around ~100 requests per second previously. This is a huge difference. But when a rotating proxy is used, the performance degrades again. HTTP connection pooling doesn't work when a rotating proxy is used and the TCP connection needs to be reestablished every time the IP changes.

**Comment by STeveShary** [comment]
> HTTP connection pooling doesn't work when a rotating proxy is used and the TCP connection needs to be reestablished every time the IP changes.

"Layer of abstraction":  The cause of and solution to all your problems!

**Comment by rehmatworks** [comment]
@STeveShary Thank you so much for guiding me in the right direction and helping me in solving this serious performance issue.
